azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -969,223 +930,336 @@ def do_retry():
-            self._clear_reconnect_timer()
-            self.send_op_down(op)
+                elif isinstance(op, pipeline_ops_base.DisconnectOperation):
+                    # First, always clear any reconnect timer. Because a manual disconnection is
+                    # ocurring, we won't want to be reconnecting any more.
+                    self._clear_reconnect_timer()
 
+                    if self.state is ReconnectState.CONNECTED:
+                        logger.debug(
+                            "{}({}): State changes CONNECTED -> DISCONNECTING. Sending op down.".format(
+                                self.name, op.name
+                            )
+                        )
+                        self.state = ReconnectState.DISCONNECTING
+                        self._add_connection_op_callback(op)
+                    elif self.state is ReconnectState.DISCONNECTED:
+                        logger.debug(
+                            "{}({}): State is already DISCONNECTED. Sending op down".format(
+                                self.name, op.name
+                            )
+                        )
+                        self._add_connection_op_callback(op)
+                    else:
+                        # This should be impossible to reach
+                        logger.warning(
+                            "{}({}): Invalid State - {}".format(self.name, op.name, self.state)
+                        )
+
+                elif isinstance(op, pipeline_ops_base.ReauthorizeConnectionOperation):
+                    if self.state is ReconnectState.CONNECTED:
+                        logger.debug(
+                            "{}({}): State changes CONNECTED -> REAUTHORIZING. Sending op down.".format(
+                                self.name, op.name
+                            )
+                        )
+                        self.state = ReconnectState.REAUTHORIZING
+                        self._add_connection_op_callback(op)
+                    elif self.state is ReconnectState.DISCONNECTED:
+                        logger.debug(
+                            "{}({}): State changes DISCONNECTED -> REAUTHORIZING. Sending op down".format(
+                                self.name, op.name
+                            )
+                        )
+                        self.state = ReconnectState.REAUTHORIZING
+                        self._add_connection_op_callback(op)
+                    else:
+                        # This should be impossible to reach
+                        logger.warning(
+                            "{}({}): Invalid State - {}".format(self.name, op.name, self.state)
+                        )
+
+                elif isinstance(op, pipeline_ops_base.ShutdownPipelineOperation):
+                    self._clear_reconnect_timer()
+
+                # In all cases the op gets sent down, but make sure to release the lock first
+                self.state_lock.release()
```

Comment:
>self.state_lock.release() [](http://example.com/codeflow?start=16&length=25)

If anything after acquire and before here raises an exception, the lock remains held.  After that, all future calls to acquire will block, and that will be disaster.  Killing the python process will be the only recovery option.